Add support for Xiaomi Mi Scale V1 and V2

tiagofreire-pt opened the pull request:
## Description:

esphome/feature-requests#209

@tsunglung have all the rights and effort for this.

**Pull request in https://github.com/esphome/feature-requests/issues/209

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs). - See: https://github.com/esphome/esphome-docs/pull/314

tiagofreire-pt committed:
Add support for Xiaomi BLE Scale V1 and V2

All the rights and effort are from @tsunglung.

tiagofreire-pt committed:
Update xiaomi_ble.h

tiagofreire-pt committed:
Create __init__.py

tiagofreire-pt committed:
Create sensor.py

tiagofreire-pt committed:
Create xiaomi_miscale.cpp

tiagofreire-pt committed:
Create xiaomi_miscale.h

OttoWinter commented on esphome/components/xiaomi_ble/xiaomi_ble.cpp:
```diff
@@ -9,6 +9,34 @@ namespace xiaomi_ble {
 static const char *TAG = "xiaomi_ble";
 
 bool parse_xiaomi_data_byte(uint8_t data_type, const uint8_t *data, uint8_t data_length, XiaomiParseResult &result) {
+  if (result.type == XiaomiParseResult::TYPE_MISCALE) {
```
Why are these in a seprate if condition? as far as I see the data_types are still unique across all xiaomi devices (you can also see this if you decompile the xiaomi apps), just put these switch cases into the switch block below.

tsunglung replied:
From https://community.home-assistant.io/t/integrating-xiaomi-mi-scale/9972/81. There is not data type from Xaiomi Scale. So I did not use switch cases with Flora. Any better idea?

OttoWinter commented on esphome/components/xiaomi_ble/xiaomi_ble.cpp:
```diff
@@ -69,40 +97,60 @@ optional<XiaomiParseResult> parse_xiaomi(const esp32_ble_tracker::ESPBTDevice &d
   }
 
-  if (!device.get_service_data_uuid()->contains(0x95, 0xFE)) {
+  if (!device.get_service_data_uuid()->contains(0x95, 0xFE) and !device.get_service_data_uuid()->contains(0x1D, 0x18) and !device.get_service_data_uuid()->contains(0x1B, 0x18))) {
```
```suggestion
  if (!device.get_service_data_uuid()->contains(0x95, 0xFE) && !device.get_service_data_uuid()->contains(0x1D, 0x18) && !device.get_service_data_uuid()->contains(0x1B, 0x18))) {
```

See also linter output, this line is too long

tsunglung replied:
OK.

OttoWinter commented on esphome/components/xiaomi_ble/xiaomi_ble.cpp:
```diff
@@ -69,40 +97,60 @@ optional<XiaomiParseResult> parse_xiaomi(const esp32_ble_tracker::ESPBTDevice &d
   bool is_mijia = (raw[1] & 0x20) == 0x20 && raw[2] == 0xAA && raw[3] == 0x01;
   bool is_miflora = (raw[1] & 0x20) == 0x20 && raw[2] == 0x98 && raw[3] == 0x00;
+  bool is_miscale =  raw[3] == 0xE3 && raw[4] == 0x07 && raw[5] == 0x08;
+  bool is_mibfs = raw[2] == 0xE3 && raw[3] == 0x07 && raw[4] == 0x08;
```
What is this Mibfs? Is it to distinguish between two different kinds of MiScale devices?

tsunglung replied:
Xiaomi Body Fat Scale which is similar to Xaiomi Scale

tiagofreire-pt replied:
One is the V1 and the other is the V2.

OttoWinter commented on esphome/components/xiaomi_ble/xiaomi_ble.cpp:
```diff
@@ -113,7 +161,7 @@ bool XiaomiListener::parse_device(const esp32_ble_tracker::ESPBTDevice &device)
     return false;
 
-  const char *name = res->type == XiaomiParseResult::TYPE_MIFLORA ? "Mi Flora" : "Mi Jia";
+  const char *name = res->type == XiaomiParseResult::TYPE_MIFLORA ? "Mi Flora" : (res->type == XiaomiParseResult::TYPE_MIJIA ? "Mi Jia": "Mi Scale");
```
This inline if is getting too long, please split it into a switch statement. (Define const char *name above, and in the switch statement set name to the correct value.)

tsunglung replied:
First, I do not know why char *name was declared as const. "char *name" was declared as const so it is read-only after assigned to value.

OttoWinter commented on esphome/components/xiaomi_miscale/__init__.py:
```diff
@@ -0,0 +1 @@
+
```
This file should be empty (no newline). Otherwise pylint will complain (see travis output)

tiagofreire-pt replied:
I can't remove the single line on the online editor.

OttoWinter commented on esphome/components/xiaomi_miscale/sensor.py:
```diff
@@ -0,0 +1,36 @@
+
+DEPENDENCIES = ['esp32_ble_tracker']
+AUTO_LOAD = ['xiaomi_ble']
+UNIT_GRAMS = 'kg'
```
1. Put these constants (UNIT_GRAMS to ICON_SCALE) into const.py - they are generic enough to be put there.

2. This UNIT_GRAMS should be called UNIT_KILOGRAMS because it is not in fact grams but kilograms.

tsunglung replied:
OK

OttoWinter commented on esphome/components/xiaomi_miscale/xiaomi_miscale.h:
```diff
@@ -0,0 +1,47 @@
+ protected:
+  uint64_t address_;
+  sensor::Sensor *weight_{nullptr};
+  sensor::Sensor *battery_level_{nullptr};
```
Battery level isn't defined in the `sensor.py` file of this PR, please either add battery level to the python codegen part or remove battery_level_ in this file.

tsunglung replied:
OK

tsunglung commented:
[xiaomi_scale.diff.zip](https://github.com/esphome/esphome/files/3498454/xiaomi_scale.diff.zip)

modified according to comments

tiagofreire-pt committed:
Update esphome/components/xiaomi_ble/xiaomi_ble.cpp

Co-Authored-By: Otto Winter <otto@otto-winter.com>

tiagofreire-pt committed:
Update __init__.py

tiagofreire-pt committed:
Update __init__.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update const.py

tiagofreire-pt committed:
Update __init__.py

tiagofreire-pt committed:
Delete __init__.py

tiagofreire-pt committed:
Create __init__.py

tiagofreire-pt commented:
@tsunglung I guess that all changes were applied. Please, for both @OttoWinter and @tsunglung if detected any inconsistency just say or suggest.

tiagofreire-pt committed:
Delete __init__.py

tiagofreire-pt committed:
Add files via upload

tiagofreire-pt committed:
Update __init__.py

tiagofreire-pt committed:
Update __init__.py

tiagofreire-pt committed:
Update sensor.py

Removing excessive empty lines.

Anonym-tsk commented:
<img width="173" alt="Home Assistant 2019-08-20 13-22-12" src="https://user-images.githubusercontent.com/1973755/63339451-94ed7980-c34d-11e9-9559-c9601f625bad.png">
Are you planning to add support for other data from mi scale?

tsunglung commented:
> <img alt="Home Assistant 2019-08-20 13-22-12" width="173" src="https://user-images.githubusercontent.com/1973755/63339451-94ed7980-c34d-11e9-9559-c9601f625bad.png">
> 
> Are you planning to add support for other data from mi scale?

Currently, no

Anonym-tsk commented:
It's better to add full support

tiagofreire-pt commented:
@Anonym-tsk, I agree with you.

Currently, this scale is basically non integrated on anything, except for some singular python scripts available on the internet.

Let's first consolidate the minimum requirements for this integration, and the next step should be increasing the data collected from the BLE stream.

tsunglung commented:
@tiagofreire-pt 
I found a mistake, please resend the commit with attached diff
[miscale.diff.zip](https://github.com/esphome/esphome/files/3526066/miscale.diff.zip)
@OttoWinter 
Can you have review it again?

tiagofreire-pt commented:
@tsunglung , I'm sorry but can you point out what was changed, please? I need to modify this by hand.

Thanks.

tsunglung commented:
[miscale.diff.zip](https://github.com/esphome/esphome/files/3526459/miscale.diff.zip)
in esphome/components/xiaomi_ble/xiaomi_ble.cpp
1. use swich-case in parse_xiaomi_data_byte() function
2. use UUID to identity mi scale type in parse_xiaomi function

in esphome/components/xiaomi_miscale/sensor.py
fix travis-ci build error

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.h

tiagofreire-pt committed:
Update const.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update xiaomi_miscale.h

tiagofreire-pt commented:
I managed to get that done.

But on the xiaomi_ble.cpp, this new block of code is not compatible with the curret status on this PR version:

`+  } else {
+    const uint8_t *data = &raw[0];
+    const uint8_t data_length = device.get_service_data().size();
+    const uint8_t raw_type = 0x16; // sdid = 22
+    result.type = XiaomiParseResult::TYPE_MISCALE;
+    success = parse_xiaomi_data_byte(raw_type, data, data_length, result);
+  }
+`

Also on sensor.py, this block ommits the units (kg + ohm):

`+DEPENDENCIES = ['esp32_ble_tracker']
+AUTO_LOAD = ['xiaomi_ble']
+
+xiaomi_miscale_ns = cg.esphome_ns.namespace('xiaomi_miscale')
+XiaomiMiscale = xiaomi_miscale_ns.class_('XiaomiMiscale', esp32_ble_tracker.ESPBTDeviceListener,
+                                         cg.Component)`

I also added some code for impedance, on many other files.

tiagofreire-pt committed:
Update const.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update const.py

tsunglung commented:
> I managed to get that done.
> 
> But on the xiaomi_ble.cpp, this new block of code is not compatible with the curret status on this PR version:
> 
> ```
> +  } else {
> +    const uint8_t *data = &raw[0];
> +    const uint8_t data_length = device.get_service_data().size();
> +    const uint8_t raw_type = 0x16; // sdid = 22
> +    result.type = XiaomiParseResult::TYPE_MISCALE;
> +    success = parse_xiaomi_data_byte(raw_type, data, data_length, result);
> +  }
> +
> ```
> 
> I also added some code for impedance, on many other files.

I attached the modified "xiaomi_ble.cpp". Does it help?
[xiaomi_ble.cpp.zip](https://github.com/esphome/esphome/files/3528004/xiaomi_ble.cpp.zip)

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

prototux commented:
> @Anonym-tsk, I agree with you.
> 
> Currently, this scale is basically non integrated on anything, except for some singular python scripts available on the internet.
> 
> Let's first consolidate the minimum requirements for this integration, and the next step should be increasing the data collected from the BLE stream.

If needed you have documentation on the mi scale v2 (aka "mibfs" and "mibcs") here: https://github.com/wiecosystem/Bluetooth/blob/master/doc/devices/huami.health.scale2.md

tiagofreire-pt committed:
Update sensor.py

$ flake8 esphome
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.ICON_FLOWER' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.CONF_TEMPERATURE' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.ICON_WATER_PERCENT' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.ICON_THERMOMETER' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.CONF_ILLUMINANCE' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.CONF_MOISTURE' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.UNIT_CELSIUS' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.CONF_CONDUCTIVITY' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.UNIT_LUX' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.ICON_BRIGHTNESS_5' imported but unused
esphome/components/xiaomi_miscale/sensor.py:4:1: F401 'esphome.const.UNIT_MICROSIEMENS_PER_CENTIMETER' imported but unused
esphome/components/xiaomi_miscale/sensor.py:14:98: E502 the backslash is redundant between brackets

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt commented:
@tsunglung, I finally solve all those warnings and errors, at least, in 2 of 4 of those Travis builds.

@OttoWinter could you review the code, please?

Thanks for both of you.

zsiti commented:
Thank you all for your great work, I'm looking forward seeing this among the officially supported devices list! :)

glmnet commented:
@tiagofreire-pt  can you check the merge conflicts?

OttoWinter commented on esphome/const.py:
```diff
@@ -470,6 +472,7 @@
 ICON_FLOWER = 'mdi:flower'
 ICON_GAS_CYLINDER = 'mdi:gas-cylinder'
 ICON_GAUGE = 'mdi:gauge'
+ICON_IMPEDANCE = 'mdi:omega'
```
```suggestion
ICON_OMEGA = 'mdi:omega'
```

Icons are always named after their icon name - they could for example be used in many different circumstances.

OttoWinter commented on esphome/const.py:
```diff
@@ -79,6 +79,7 @@
 CONF_CODE = 'code'
 CONF_COLD_WHITE = 'cold_white'
 CONF_COLD_WHITE_COLOR_TEMPERATURE = 'cold_white_color_temperature'
+CONF_WEIGHT = 'weight'
```
Keep this list alphabetically sorted please.

tiagofreire-pt replied:
Done.

OttoWinter commented on esphome/const.py:
```diff
@@ -502,6 +505,7 @@
 UNIT_HZ = 'hz'
 UNIT_HECTOPASCAL = 'hPa'
 UNIT_KELVIN = 'K'
+UNIT_KILOGRAMS = 'kg'
```
```suggestion
UNIT_KILOGRAM = 'kg'
```

The unit names are all singular.

OttoWinter reviewed (commented):
Thanks, some last cleanup left, and then it should be fine.

@glmnet is right, you need to rebase this branch on top of dev again to make it merge-able with the latest xiaomi merges.

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update esphome/const.py

Co-Authored-By: Otto Winter <otto@otto-winter.com>

tiagofreire-pt committed:
Update esphome/const.py

Co-Authored-By: Otto Winter <otto@otto-winter.com>

tiagofreire-pt commented on esphome/components/xiaomi_miscale/sensor.py:
```diff
@@ -0,0 +1,46 @@
+    CONF_WEIGHT, UNIT_KILOGRAMS, ICON_SCALE, \
+    UNIT_OHM, CONF_IMPEDANCE, ICON_IMPEDANCE
+
+# from esphome.const import CONF_BATTERY_LEVEL, CONF_MAC_ADDRESS, CONF_TEMPERATURE, \
```
Done.

tiagofreire-pt committed:
Update xiaomi_ble.h

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt commented:
@OttoWinter, I did the chances on .cpp and .h for BLE.

OttoWinter commented:
@tiagofreire-pt Great, thanks! Only thing left is to rebase the branch on top of the latest dev branch, see contributing guide for instructions.

tiagofreire-pt commented:
I see on https://esphome.io/guides/contributing.html instructions for "git" program but not how to do it straight on gibhub.com.

OttoWinter commented:
@tiagofreire-pt I believe that is not possible from the GitHub UI (maybe there's a Resolve Conflicts button somewhere, dunno). You will probably have to install git on your machine and fix it that way. Or (a hackier solution), close the PR and create a new PR from dev.

tiagofreire-pt committed:
Merge branch 'dev' into patch-1

tiagofreire-pt commented:
> @tiagofreire-pt I believe that is not possible from the GitHub UI (maybe there's a Resolve Conflicts button somewhere, dunno). You will probably have to install git on your machine and fix it that way. Or (a hackier solution), close the PR and create a new PR from dev.

Done.

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

OttoWinter commented:
@tiagofreire-pt Looks like the merge did not work completely: See https://travis-ci.org/esphome/esphome/jobs/579231824#L951

tiagofreire-pt commented:
> @tiagofreire-pt Looks like the merge did not work completely: See https://travis-ci.org/esphome/esphome/jobs/579231824#L951

Yes, it seems to lack of some declaration of variables. @tsunglung, could you help, please?

tsunglung commented:
[xiaomi_ble.cpp.zip](https://github.com/esphome/esphome/files/3563046/xiaomi_ble.cpp.zip)
@tiagofreire-pt when you rebase, you have a mistake. I corrected it and attached the modified file.

tiagofreire-pt commented:
> 
> 
> [xiaomi_ble.cpp.zip](https://github.com/esphome/esphome/files/3563046/xiaomi_ble.cpp.zip)
> @tiagofreire-pt when you rebase, you have a mistake. I corrected it and attached the modified file.

Thanks, @tsunglung. 

But I can't find the difference in the code between the one to be merged and the one that you attanched (xiami_ble.cpp.zip). They are exactly the same, ipsis verbis. Coul you check that, please?

Thanks.

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tsunglung commented:
> > [xiaomi_ble.cpp.zip](https://github.com/esphome/esphome/files/3563046/xiaomi_ble.cpp.zip)
> > @tiagofreire-pt when you rebase, you have a mistake. I corrected it and attached the modified file.
> 
> Thanks, @tsunglung.
> 
> But I can't find the difference in the code between [the one to be merged](https://github.com/esphome/esphome/blob/f549cce6bf59b86f2a678c60254cf0db9c396056/esphome/components/xiaomi_ble/xiaomi_ble.cpp) and the one that you just attached (xiami_ble.cpp.zip). They are exactly the same, _ipsis verbis_. Could you check that, please?
> 
> Thanks.

OK, I upload again
[xiaomi_ble.cpp.zip](https://github.com/esphome/esphome/files/3621186/xiaomi_ble.cpp.zip)

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt commented:
@tsunglung, I did everything, and can't understand why this code is currently crashing in the last test. Could you provide some debugging help on this, please?

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_miscale.h

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt commented:
Finally, after lots of indentation errors and formatting mishaps, the code gets a green light. Yay! 👍

tiagofreire-pt commented:
@OttoWinter , could you merged it, please? :)

tsunglung commented:
> Finally, after lots of indentation errors and formatting mishaps, the code gets a green light. Yay! 👍

Great!, I tested your code. No crash found.

askthisthing commented:
@tsunglung @tiagofreire-pt Nice job, thank you! Could you please show me what to put into the YAML config to make my MiScale V1 talk to ESPHome (ESP32)? I use docker and I don't know how to build/flash/test this code, so I thought I'd wait until it gets merged into the dev branch, and the official docker image gets auto-released.
@OttoWinter do you think you could merge the code? Many thanks for your awesome work!

glmnet commented:
Hey there, docs needs more work

ned-kelly commented:
> Hey there, docs needs more work

Hi @glmnet 

Love to see this one in Master - What documentation is required (Quick bullet point list is sufficient) and I might crack on with this over the weekend so we can push this Merge live.

Cheers,

tiagofreire-pt commented:
> @tsunglung @tiagofreire-pt Nice job, thank you! Could you please show me what to put into the YAML config to make my MiScale V1 talk to ESPHome (ESP32)? 

Thanks, @askthisthing.

Please, check this file: https://github.com/esphome/esphome-docs/pull/314/files#diff-d5d905c0eaebc87093babfd9e962674a

askthisthing commented:
@tiagofreire-pt Thank you! I could prepare the YAML, so now I'll just have to wait until this PR gets merged, and we could use it officially 😃

pickeld commented:
Cant wait! Any idea when it will be released?
Thank you all for your time!

diraimondo commented:
It would be necessary to fix the PR on the documentation but, in any case, we have to wait the return of @OttoWinter. It looks busy in this period.

OttoWinter committed:
Merge branch 'dev' into xiaomi-scale

# Conflicts:
#	esphome/components/xiaomi_ble/xiaomi_ble.cpp
#	esphome/components/xiaomi_ble/xiaomi_ble.h
#	esphome/const.py

OttoWinter committed:
Update xiaomi_ble.cpp

OttoWinter commented:
Update: I just updated this PR to remove the conflicts, but there's one thing left. In https://github.com/esphome/esphome/pull/755, we renamed all xiaomi integrations to the model numbers of the devices to make things more clear. I would like to do the same for this integration.

Does someone have the model numbers of the Xiaomi Scale? I can't find any info on the xiaomi website.

lupin-de-mid commented:
My device has Bluetooth le name MI_SCALE and on bottom something like `xmtzc01hm`
According to internet (  [russian](url) ) mi sclae 2 (2019) has name  `xmtzc04hm`

Alex9779 commented:
1. Name: XIAOMI Waage 2.0
2. Modell: C98167

Alex9779 commented:
But its hard... i think the number of lupin are good...

OttoWinter committed:
Merge branch 'dev' into xiaomi-scale

OttoWinter committed:
Rename to XMTZC0XHM

pplucky commented:
> Update: I just updated this PR to remove the conflicts, but there's one thing left. In #755, we renamed all xiaomi integrations to the model numbers of the devices to make things more clear. I would like to do the same for this integration.
> 
> Does someone have the model numbers of the Xiaomi Scale? I can't find any info on the xiaomi website.

Mine shows in lescan as MIBFS it is named 'Mi Body Composition Scale 2' and model number is XMTZC05HM
![IMG_2066](https://user-images.githubusercontent.com/16407309/67245321-9887a480-f453-11e9-9e59-a22c8c79d6af.jpg)

bpmittal commented:
![IMG-20191025-WA0000](https://user-images.githubusercontent.com/45149160/67517183-a2addb00-f6bf-11e9-9d68-8423c48a069d.jpg)

i have this model available in India,

pickeld commented:
so its basically done. what's left ? when will it be published.  cant wait.

usenetj23 commented:
Any news when this will be released?

zsiti commented:
> My device has Bluetooth le name MI_SCALE and on bottom something like `xmtzc01hm`
> According to internet ( [russian](url) ) mi sclae 2 (2019) has name `xmtzc04hm`

I can confirm what @lupin-de-mid says, my device also has a bluetooth name MI_SCALE, and on it's bottom side the model name is XMTZC01HM.
![miscale-model-name](https://user-images.githubusercontent.com/13167838/68093252-e78dfa80-fe93-11e9-99bc-5da08ae3bc3d.jpg)

askthisthing commented:
I'm very much looking forward to the Mi Scale integration, I hope it gets released soon. Do we have an (approximate) ETA on this?

tiagofreire-pt commented:
My model is XMTZC0**2**HM, with FCC ID: 2AC8UXMTZC02HM.

askthisthing commented:
I suppose we have all the model numbers gathered, is it much work to implement those in this code? Can't wait for this release :)

tiagofreire-pt committed:
esphome/components/xiaomi_xmtzc0xhm/sensor.py:15:42: E128 continuation line under-indented for visual indent

tiagofreire-pt commented:
@tsunglung, would you help here on solving the compiling errors, please? Thanks!

pickeld commented:
any update here?

askthisthing commented:
Hi, can we please get an update if this is on the roadmap? We've been waiting for months, even though that by the looks of it the feature is done. Thanks in advance!

tiagofreire-pt commented:
> 
> 
> Hi, can we please get an update if this is on the roadmap? We've been waiting for months, even though that by the looks of it the feature is done. Thanks in advance!

Yap, you're right. But this requires some code revision as the ESPHome renamed a lot of Xiaomi platform devices. I believe @tsunglung and @OttoWinter could give a hand on this, as more and more users are willing to have this implemented on production core. :)

I'm also available to help, as far as my skills are enough for. ;)

askthisthing commented:
@tiagofreire-pt Thanks for the answer.

> Yap, you're right. But this requires some code revision as the ESPHome renamed a lot of Xiaomi platform devices. I believe @tsunglung and @OttoWinter could give a hand on this, as more and more users are willing to have this implemented on production core. :)

Sure, I understand, and I actually support the idea of consistent naming convention for Xiaomi devices, but @OttoWinter's [last comment was on 2019-Oct-17, asking users to share their Mi Scale device model numbers](https://github.com/esphome/esphome/pull/691#issuecomment-543334323), answers came in, but no status update from Otto yet in spite of the numerous and periodic status questions from users. No hate of course, this is a free open source project, and if I'd have the skills, I'd gladly help out, but also t's unclear for me (and seemingly other users as well) what's the blocker on this issue since we have the MiScale model numbers and naming. Most of other similar projects like OpenMQTTgateway have successfully implemented this feature many months ago, but of course I'd like to keep using ESPHome, as it's far easier to configure and implement - and also my other devices rely on ESPHome as well.
Thanks for the great project, and hope we can see MiScale integration in the near future :)

panuruj committed:
Split the SGP30 baseline into 2 values

- According to the SGP30 datasheet, each eCO2 and TVOC baseline is a 2-byte value (MSB first)
- The current implementation ignores the MSB of each of the value
- Update the schema to allow 2 different baseline values (optional, but both need to be specified for the baseline to apply)

panuruj committed:
Make both eCO2 and TVOC required if the optional baseline is defined

panuruj committed:
Make dump_config() looks better

HydrelioxGitHub commented:
Any update on this ?

timsavage committed:
Use b''.decode() instead of str(b'')

Handling of request arguments in WizardRequestHandler is not decoding
bytes and rather just doing a str conversion resulting in a value of
"b''" being supplied to the wizard code.

Valcob committed:
Adding the espressif 2.6.3

Commit:
add miscale

Commit:
Merge branch 'dev' of github.com:dmkif/esphome; commit 'refs/pull/936/head' of github.com:esphome/esphome into dev

Commit:
fix hdc1080 timing problem

Commit:
Merge commit 'refs/pull/944/head' of github.com:esphome/esphome into dev

Commit:
Merge commit 'refs/pull/941/head' of github.com:esphome/esphome into dev

Commit:
Merge branch 'dev' of github.com:esphome/esphome into dev

Commit:
add wakeup pin

zsiti commented:
Hi everyone! This one seemed to be a popular request, let's not leave it fade into oblivion :) @OttoWinter can you chime in please? Thank you!

Commit:
Merge remote-tracking branch 'origin' into ccs8811

Commit:
Merge branch 'dev', remote-tracking branch 'origin' into miscale

Commit:
Merge branch 'ccs8811' into miscale

Commit:
add miscale v2 support

pickeld commented:
meanwhile, if anyone wants to integrate the mi scale into HASS, have a look at this:
https://github.com/lolouk44/xiaomi_mi_scale

Commit:
Merge branch 'dev' of github.com:esphome/esphome into miscale

Commit:
Merge branch 'dev' of github.com:esphome/esphome into miscale

Commit:
Merge branch 'dev' of github.com:esphome/esphome into miscale

Commit:
Merge branch 'dev' of github.com:esphome/esphome into miscale

Commit:
Merge branch 'dev' of github.com:esphome/esphome into miscale

Commit:
Merge branch 'dev' of github.com:esphome/esphome into miscale

llonca13 commented:
What happened :) ?

jan-loeffler commented:
I tested the code in tiagofreire-pt:patch-1 and got it to work with minor changes (needed to update platformio and fixed a small error:
`compilation terminated.
*** [.pioenvs/ble_gateway/src/esphome/components/wifi/wifi_component.cpp.o] Error 1
src/esphome/components/xiaomi_ble/xiaomi_ble.cpp: In function 'esphome::optional<esphome::xiaomi_ble::XiaomiParseResult> esphome::xiaomi_ble::parse_xiaomi(const esphome::esp32_ble_tracker::ESPBTDevice&)':
src/esphome/components/xiaomi_ble/xiaomi_ble.cpp:131:7: error: 'is_mijia' was not declared in this scope
   if (is_mijia || is_miflora || is_lywsd02 || is_cgg1) {
       ^
src/esphome/components/xiaomi_ble/xiaomi_ble.cpp:131:19: error: 'is_miflora' was not declared in this scope
   if (is_mijia || is_miflora || is_lywsd02 || is_cgg1) {`

After that i got it working with my mi scale v1 
![IMG_20200610_132445](https://user-images.githubusercontent.com/8266989/84262701-5912e300-ab1e-11ea-841f-2328b4aa1821.jpg)
![Annotation 2020-06-10 133309](https://user-images.githubusercontent.com/8266989/84263161-4351ed80-ab1f-11ea-85a3-2ab0f82d521e.png)
![Annotation 2020-06-10 133452](https://user-images.githubusercontent.com/8266989/84263167-43ea8400-ab1f-11ea-962a-2bc1f022b642.png)

My config:

```
esphome:
  name: ble_gateway
  platform: ESP32
  board: esp-wrover-kit

wifi:
  ssid: "Wifi Name"
  password: "Simple Password"

ap:
    ssid: "Ble Gateway Fallback Hotspot"
    password: "Even more simple Password"

captive_portal:

logger:

api:
web_server:
  port: 80
ota:
esp32_ble_tracker:
sensor:
    -   platform: xiaomi_xmtzc0xhm
        mac_address: 'C8:47:00:00:00:00'
        weight:
          name: "Xiaomi Mi Scale Weight"
```

It also works out of the box in Home Assistant:
![Annotation 2020-06-10 133844](https://user-images.githubusercontent.com/8266989/84263422-c8d59d80-ab1f-11ea-829e-41b99e759058.png)

tiagofreire-pt committed:
Merge branch 'patch-1' into miscale

tiagofreire-pt committed:
Merge pull request #1 from dmkif/miscale

Miscale

tiagofreire-pt committed:
Merge branch 'dev' into patch-1

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update xiaomi_ble.h

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update const.py

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update const.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update sensor.py

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update ccs811.cpp

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update ccs811.h

tiagofreire-pt committed:
Update ccs811.cpp

tiagofreire-pt committed:
Update ccs811.cpp

tiagofreire-pt committed:
Update ccs811.cpp

tiagofreire-pt committed:
Update ccs811.h

tiagofreire-pt committed:
Update xiaomi_ble.cpp

tiagofreire-pt committed:
Update xiaomi_ble.h

tiagofreire-pt committed:
Update ccs811.h

tiagofreire-pt committed:
Update xiaomi_ble.cpp

jan-loeffler commented:
> @OttoWinter , could you check why the CI is not running the tests at this moment, please?
> 
> Thank you!

Is there any problem, that needs additional work / help?

glmnet commented:
@tiagofreire-pt I run travis again, is failing, would you like to check it?

I'd love to have @ahpohl review this one if possible.

Commit:
Smaller Fixes

jan-loeffler commented:
I made some small changes: https://github.com/tiagofreire-pt/esphome/pull/2 
With the changes it should run fine @tiagofreire-pt  can you check and confirm?

tiagofreire-pt committed:
Merge pull request #2 from jan-loeffler/patch-1

Smaller Fixes

tiagofreire-pt commented:
@jan-loeffler, thanks for the fixes! 🥇 

Would you mind checking why is still failing the tests, please?

Thanks again!

Commit:
Clang Formating

jan-loeffler commented:
I just run the Clang Formater https://github.com/tiagofreire-pt/esphome/pull/3

tiagofreire-pt committed:
Merge pull request #3 from jan-loeffler/patch-1

Clang Formating

jan-loeffler commented:
Now just one test is failing, as far as i got it it is mainly down to warnings

pattyland commented:
@jan-loeffler So there seems to be one last error with an undefined reference for every xiami component if i got this right. I'm having a hard time trying to understand what's wrong:

```log
.pioenvs/test2/src/esphome/components/xiaomi_wx08zm/xiaomi_wx08zm.cpp.o: In function `esphome::xiaomi_wx08zm::XiaomiWX08ZM::parse_device(esphome::esp32_ble_tracker::ESPBTDevice const&)':
/home/travis/build/esphome/esphome/tests/build/test2/src/esphome/components/xiaomi_wx08zm/xiaomi_wx08zm.cpp:27: undefined reference to `esphome::xiaomi_ble::parse_xiaomi_header(esphome::esp32_ble_tracker::ServiceData const&)'
/home/travis/build/esphome/esphome/tests/build/test2/src/esphome/components/xiaomi_wx08zm/xiaomi_wx08zm.cpp:38: undefined reference to `esphome::xiaomi_ble::parse_xiaomi_message(std::vector<unsigned char, std::allocator<unsigned char> > const&, esphome::xiaomi_ble::XiaomiParseResult&)'
/home/travis/build/esphome/esphome/tests/build/test2/src/esphome/components/xiaomi_wx08zm/xiaomi_wx08zm.cpp:41: undefined reference to `esphome::xiaomi_ble::report_xiaomi_results(esphome::optional<esphome::xiaomi_ble::XiaomiParseResult> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
*** [.pioenvs/test2/firmware.elf] Error 1
```

It would be so awesome to get this in 1.15...

jan-loeffler commented:
@pattyland Sorry i don't have time at the moment

nldroid commented:
Any news on this? Anything that can be done?

dckiller51 commented:
> des nouvelles à ce sujet? Tout ce qui peut être fait?

Also waiting for this integration. In order to migrate BT-MQTT-GATEWAY

dckiller51 commented:
Still not present in the new update. Too bad.

jesserockz reviewed (changes requested):
Please rebase this branch from the latest dev as there have been significant changes to the `xiaomi_ble` classes and it conflicts a lot now.

dckiller51 commented:
https://github.com/esphome/esphome/pull/1368

dckiller51 commented:
> @ tiagofreire-pt Y a - t-il une chance que cela soit bientôt rebasé? J'ai essayé mais je ne peux pas comprendre tout le rebasage. Ce serait bien de l'inclure dans l'esphome.

hello, you can try my version. it works. Pull 1368

nickrout commented:
Actually @dckiller51 I saw your pull right after I posted, thanks I'll try and get time tonight to try it.

dckiller51 commented:
> En fait @ dckiller51 J'ai vu votre pull juste après avoir posté, merci je vais essayer d'avoir le temps ce soir pour l'essayer.



> En fait @ dckiller51 J'ai vu votre pull juste après avoir posté, merci je vais essayer d'avoir le temps ce soir pour l'essayer.

if you have a miscale2 can test with xiaomi_xmtzc0xhm it is a version with only one directory. With my miscale 1 it works but I would like a valid person with the 2. That way I could push the version to the dev so that he validates to integrate it definitively. Thank you

makefu commented:
everything in this PR looks super promising and i'd love to see this being merged into esphome!
Cheers and Thanks!

dckiller51 commented:
> tout dans ce PR semble super prometteur et j'aimerais voir cela fusionné dans un esphome!
> Bravo et merci!
You can test this version. miscale without impedance otherwise the config below````
sensor:
  - platform: xiaomi_xmtzc0xhm
    mac_address: 'XX:XX:XX:XX:XX:XX'
    weight:
      name: "Xiaomi Mi Scale Weight"
    impedance:
      name: "Xiaomi Mi Scale Impedance"````
https://github.com/esphome/esphome/pull/1368
